pysteps/downscaling/rainfarm.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+"""
+pysteps.downscaling.rainfarm
+============================
+
+Implementation of the RainFARM stochastic downscaling method as described in
+:cite:`Rebora2006`.
+
+.. autosummary::
+    :toctree: ../generated/
+
+    downscale
+"""
+
+import warnings
+
+import numpy as np
+from scipy.ndimage import convolve
+
+
+def _log_slope(log_k, log_power_spectrum):
+    lk_min = log_k.min()
+    lk_max = log_k.max()
+    lk_range = lk_max - lk_min
+    lk_min += (1 / 6) * lk_range
+    lk_max -= (1 / 6) * lk_range
+    dk = lk_range / 32
+
+    selected = (lk_min <= log_k) & (log_k <= lk_max)
+    num_selected = np.count_nonzero(selected)
+    lk_sel = log_k[selected]
+    ps_sel = log_power_spectrum[selected]
+    (alpha, _) = np.polyfit(lk_sel, ps_sel, 1)
+    alpha = -alpha
+
+    return alpha
+
+
+def _balanced_spatial_average(x, k):
+    ones = np.ones_like(x)
+    return convolve(x, k) / convolve(ones, k)
+
+
+def downscale(P, alpha=None, ds_factor=16, threshold=None):
+    """
+    Downscale a rainfall field by a given factor.
+
+    Parameters
+    ----------
+
+    P : array_like
+        Array of shape (m,n) containing the input field.
+        The input is expected to contain rain rate values.
+
+    alpha : float, optional
+        Spectral slope. If none, the slope is estimated from
+        the input array.
+
+    ds_factor : int, optional
+        Downscaling factor.
+
+    threshold : float, optional
+        Set all values lower than the threshold to zero.
+
+
+    Returns
+    -------
+    r : array_like
+        Array of shape (m*ds_factor,n*ds_factor) containing
+        the downscaled field.
+
+    Notes
+    -----
+    The pysteps implementation of RainFARM currently covers spatial downscaling
+    only, that is, it can improve spatial resolution of a rainfall field, but
+    unlikely the original algorithm from Rebora et al. (2006), it cannot downscale
+    the temporal dimension.
```

Comment:
```suggestion
    Currently, the pysteps implementation of RainFARM only covers spatial downscaling.
    That is, it can improve the spatial resolution of a rainfall field. However, unlike
    the original algorithm from Rebora et al. (2006), it cannot downscale the temporal
    dimension.
```